src/replication/replication_manager.cpp

```diff
@@ -0,0 +1,290 @@
+      }
+      lock.unlock();
+    }
+  } catch (const MessengerException &e) {
+    REPLICATION_LOG_WARN(fmt::format("[FAILED] Send -> {} (type {} block {}): msg size {}", replica_name,
+                                     static_cast<uint8_t>(type), block, msg.size()));
+  }
+}
+
+void ReplicationManager::ReplicateBuffer(storage::BufferedLogWriter *buffer) {
+  if (!replication_enabled_) {
+    REPLICATION_LOG_WARN(fmt::format("Skipping replicate buffer as replication is disabled."));
+    return;
+  }
+
+  NOISEPAGE_ASSERT(buffer != nullptr,
+                   "Don't try to replicate null buffers. That's pointless."
+                   "You might plausibly want to track statistics at some point, but that should not happen here.");
+  // TODO(WAN): Is it true that a replica will never want to send its buffers to the primary?
+  // TODO(WAN): Cache the identity check into a bool?
+  if (IsPrimary()) {
+    common::json j;
+    // TODO(WAN): Add a size and checksum to message.
+    j["buf_id"] = next_buffer_sent_id_++;
+    j["content"] = nlohmann::json::to_cbor(std::string(buffer->buffer_, buffer->buffer_size_));
+
+    for (const auto &replica : replicas_) {
+      // TODO(WAN): many things break when block is flipped from true to false.
+      ReplicaSend(replica.first, MessageType::REPLICATE_BUFFER, j.dump(), true);
+    }
+  }
+
+  if (buffer->MarkSerialized()) {
+    empty_buffer_queue_->Enqueue(buffer);
+  }
+}
+
+void ReplicationManager::EventLoop(common::ManagedPointer<messenger::Messenger> messenger,
+                                   const messenger::ZmqMessage &msg) {
+  switch (static_cast<MessageType>(msg.GetDestinationCallbackId())) {
+    case MessageType::ACK: {
+      nlohmann::json json = nlohmann::json::parse(msg.GetMessage());
+      REPLICATION_LOG_TRACE(fmt::format("ACK: {}", json.dump()));
+      auto callback_id = json.at("callback_id").get<uint64_t>();
+      auto callback = messenger->GetCallback(callback_id);
+      (*callback)(messenger, msg);
+      break;
+    }
+    case MessageType::HEARTBEAT:
+      REPLICATION_LOG_TRACE(fmt::format("Heartbeat from: {}", msg.GetRoutingId()));
+      break;
+    case MessageType::REPLICATE_BUFFER: {
+      // Parse the buffer from the received message and add it to the provided replication logs.
+      {
+        nlohmann::json message = nlohmann::json::parse(msg.GetMessage());
+        // TODO(WAN): Sanity-check the received message.
+
+        uint64_t source_callback_id = msg.GetSourceCallbackId();
+        uint64_t msg_id = message.at("buf_id");
+        REPLICATION_LOG_TRACE(fmt::format("ReplicateBuffer from: {} {}", msg.GetRoutingId(), msg_id));
+
+        // Check if the message needs to be buffered.
+        if (msg_id > last_record_received_id_ + 1) {
+          // The message should be buffered if there are gaps in between the last seen buffer.
+          received_message_queue_.push(msg);
```

Comment:
Why not push the parsed `message` to avoid parsing it again?

Reply:
yeah, you're right. The reason was that multiple fields were being read (msg_id, content). I could create a struct for all the fields that matter.

Reply:
Refactored to do the struct, thoughts?

Reply:
I like it